### PR TITLE
Adds support for "pluggable" architecture in //rocks package

### PR DESCRIPTION
**Pull request by lolski**
## What is the goal of this PR?

We have added support for pluggable architecture in Core. The primary goal is to allow Cluster to supply its own replacement for `RocksDatabase`, `RocksTransaction`, and `RocksStorage`.

## What are the changes implemented in this PR?

- Modify the implementation of "rocks" classes such that class instantiations are always done through "creator methods" in `RocksCreator`. 
- Add `RocksCreator` implementation of Core - which returns the appropriate "rocks" objects. In Cluster, it will have a different `RocksCreator` which returns rocks objects that supports replication
- Made fields in `RocksDatabase`, `RocksTransaction`, and `RocksStorage` protected, to allow subclassing.

### CODE REVIEW

File: rocks/RocksCreator.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
+package grakn.core.rocks;
+
+import grakn.core.common.parameters.Arguments;
+import grakn.core.common.parameters.Options;
+
+import java.nio.file.Path;
+
+public class RocksCreator {
```
Comment (lolski) [inline rocks/RocksCreator.java:8]: `RocksCreator` is responsible for creating "rocks" instances (database manager, database, session, transaction, and storage). It is a "factory" class that can create either Core or Cluster specific rocks instances.

For example, the Core implementation of `RocksCreator` here will return a `RocksDatabase` that is not replicated. The `RockCreator` in Cluster however, will return a `RocksDatabase` that supports replication.

Comment (lolski) [reply rocks/RocksCreator.java:8]: Following the factory approach allows for Grakn Cluster to reuse Core's `//rocks` package with minimal modification - only re-creating the "rocks" classes that need to be customised.

File: rocks/RocksDatabase.java [hunk 2]
```diff
@@ -56,22 +56,24 @@
 
 public class RocksDatabase implements Grakn.Database {
```
Comment (lolski) [inline rocks/RocksDatabase.java:57]: `RocksDatabase` needs to open up fields as needed by the subclass in Cluster.

File: rocks/RocksStorage.java [hunk 3]
```diff
@@ -43,7 +43,7 @@
 import static grakn.core.common.collection.Bytes.bytesHavePrefix;
 import static grakn.core.common.exception.ErrorMessage.Transaction.TRANSACTION_CLOSED;
 
-class RocksStorage implements Storage {
+public class RocksStorage implements Storage {
```
Comment (lolski) [inline rocks/RocksStorage.java:46]: Open up fields and class modifier as needed by the subclass in Cluster.

File: rocks/RocksTransaction.java [hunk 4]
```diff
@@ -46,16 +46,16 @@
 
 public abstract class RocksTransaction implements Grakn.Transaction {
 
-    final RocksSession session;
-    final Arguments.Transaction.Type type;
+    protected final RocksSession session;
```
Comment (lolski) [inline rocks/RocksTransaction.java:49]: Open up fields as needed by the subclass in Cluster.

File: rocks/RocksCreator.java [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
```
[the remark this answers is not in the document]
Comment (haikalpribadi) [reply rocks/RocksCreator.java:8]: We may need multiple `Factory` classes, in which all can be held under a `Factories` class.

File: rocks/RocksStorage.java [hunk 6]
```diff
@@ -56,9 +56,9 @@
     private final Snapshot snapshot;
     private final ManagedReadWriteLock readWriteLock;
     private final AtomicBoolean isOpen;
-    final Transaction rocksTx;
+    protected final Transaction rocksTx;
```
Comment (haikalpribadi) [inline rocks/RocksStorage.java:59]: Let's rename this to `storageTransaction`

**Review by haikalpribadi — CHANGES_REQUESTED** [review]

File: rocks/RocksDatabase.java [hunk 7]
```diff
@@ -56,65 +56,78 @@
...
 
-    private final String name;
-    private final RocksGrakn rocksGrakn;
-    private final OptimisticTransactionDB rocksData;
-    private final OptimisticTransactionDB rocksSchema;
-    private final KeyGenerator.Data.Persisted dataKeyGenerator;
+    private final RocksGrakn grakn;
```
Comment (lolski) [inline rocks/RocksDatabase.java:59]: group fields based on who belong together

File: rocks/RocksGrakn.java [hunk 8]
```diff
@@ -47,31 +47,35 @@
...
+        return open(directory, new Options.Database(), new RocksFactory());
+    }
+
+    public static RocksGrakn open(Path directory, Factory factory) {
+        return open(directory, new Options.Database(), factory);
     }
 
-    public static RocksGrakn open(Path directory, Options.Database options) {
-        return new RocksGrakn(directory, options);
+    public static RocksGrakn open(Path directory, Options.Database options, Factory factory) {
```
Comment (haikalpribadi) [inline rocks/RocksGrakn.java:74]: `databaseFactory`

**Review by haikalpribadi — APPROVED** [review]
Approving, but see the comments below.

File: rocks/RocksDatabase.java [hunk 9]
```diff
@@ -56,65 +56,78 @@
...
         schemaKeyGenerator = new KeyGenerator.Schema.Persisted();
         dataKeyGenerator = new KeyGenerator.Data.Persisted();
         sessions = new ConcurrentHashMap<>();
         dataWriteSchemaLock = new StampedLock();
 
         try {
-            rocksSchema = OptimisticTransactionDB.open(this.rocksGrakn.rocksOptions(), directory().resolve(Encoding.ROCKS_SCHEMA).toString());
-            rocksData = OptimisticTransactionDB.open(this.rocksGrakn.rocksOptions(), directory().resolve(Encoding.ROCKS_DATA).toString());
+            rocksSchema = OptimisticTransactionDB.open(this.grakn.rocksOptions(), directory().resolve(Encoding.ROCKS_SCHEMA).toString());
+            rocksData = OptimisticTransactionDB.open(this.grakn.rocksOptions(), directory().resolve(Encoding.ROCKS_DATA).toString());
```
[the remark this answers is not in the document]
Comment (haikalpribadi) [reply rocks/RocksDatabase.java]: actually, let's keep it `grakn`

File: rocks/RocksGrakn.java (hunk 8 above)
Comment (lolski) [reply rocks/RocksGrakn.java]: Will do. It should be `graknFactory` to be precise (look at the type, it is `Factory` not `Factory.Database`)

**Pull request merged**